// ngx-progressive-image-loader/esm5/lib/picture-source/picture-source.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Inject, Injector, Input, Renderer2 } from '@angular/core';
import { WINDOW } from 'ngx-window-token';
var PictureSourceDirective = /** @class */ (function () {
    function PictureSourceDirective(_ElementRef, _Renderer, window, _Injector) {
        this._ElementRef = _ElementRef;
        this._Renderer = _Renderer;
        this.window = window;
        this._Injector = _Injector;
    }
    /**
     * @return {?}
     */
    PictureSourceDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // this.imageElement = this._ElementRef.nativeElement;
        // if (isSupportIntersectionObserver(this.window) && !isSpider(this.window)) {
        //   this.setOnLoadedAction();
        // } else {
        //   // show image directly
        //   loadImage(this._Renderer, this.imageElement);
        // }
        // this.setDataSrc();
        // this.ProgressiveImageLoader = this._Injector.get(ProgressiveImageLoaderComponent);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    PictureSourceDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!changes["src"].isFirstChange()) {
            this.setDataSrc();
            this.ProgressiveImageLoader.intersectionObserver.observe(this.imageElement);
        }
    };
    /**
     * @return {?}
     */
    PictureSourceDirective.prototype.setDataSrc = /**
     * @return {?}
     */
    function () {
        this._Renderer.setAttribute(this.imageElement, 'data-src', this.src);
        this.srcset && this._Renderer.setAttribute(this.imageElement, 'data-srcset', this.srcset);
    };
    PictureSourceDirective.decorators = [
        { type: Directive, args: [{
                    selector: 'source[ngxPictureSource], img[ngxPictureSource]'
                },] }
    ];
    /** @nocollapse */
    PictureSourceDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: undefined, decorators: [{ type: Inject, args: [WINDOW,] }] },
        { type: Injector }
    ]; };
    PictureSourceDirective.propDecorators = {
        src: [{ type: Input }],
        srcset: [{ type: Input }]
    };
    return PictureSourceDirective;
}());
export { PictureSourceDirective };
if (false) {
    /** @type {?} */
    PictureSourceDirective.prototype.src;
    /** @type {?} */
    PictureSourceDirective.prototype.srcset;
    /** @type {?} */
    PictureSourceDirective.prototype.imageElement;
    /** @type {?} */
    PictureSourceDirective.prototype.ProgressiveImageLoader;
    /** @type {?} */
    PictureSourceDirective.prototype._ElementRef;
    /** @type {?} */
    PictureSourceDirective.prototype._Renderer;
    /** @type {?} */
    PictureSourceDirective.prototype.window;
    /** @type {?} */
    PictureSourceDirective.prototype._Injector;
}
