// ngx-progressive-image-loader/esm5/lib/config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function IImageLoaderOptions() { }
if (false) {
    /** @type {?|undefined} */
    IImageLoaderOptions.prototype.placeholderImageSrc;
    /** @type {?} */
    IImageLoaderOptions.prototype.imageRatio;
    /** @type {?|undefined} */
    IImageLoaderOptions.prototype.filter;
    /** @type {?|undefined} */
    IImageLoaderOptions.prototype.isAggressiveLoading;
    /** @type {?|undefined} */
    IImageLoaderOptions.prototype.concurrentLoading;
}
/** @type {?} */
export var IMAGE_LOADER_CONFIG_TOKEN = new InjectionToken('Image loader Config');
/** @type {?} */
export var DEFAULT_IMAGE_LOADER_OPTIONS = (/** @type {?} */ ({
    // root?: Element | null;
    rootMargin: '10px',
    threshold: [0.1, 0.5, 1],
    imageRatio: 16 / 9,
    placeholderImageSrc: '',
    isAggressiveLoading: true,
    concurrentLoading: 4
}));
