// ngx-progressive-image-loader/esm5/lib/ngx-progressive-image-loader.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { IMAGE_LOADER_CONFIG_TOKEN } from './config';
import { ConfigurationService } from './configuration.service';
import { ImagePlaceholderComponent } from './image-placeholder/image-placeholder.component';
import { ProgressiveImageLoaderComponent } from './progressive-image-loader/progressive-image-loader.component';
import { ProgressiveImageDirective } from './progressive-image/progressive-image.directive';
/**
 * @param {?} options
 * @return {?}
 */
export function ConfigurationServiceFactory(options) {
    return new ConfigurationService(options);
}
var NgxProgressiveImageLoaderModule = /** @class */ (function () {
    function NgxProgressiveImageLoaderModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    NgxProgressiveImageLoaderModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgxProgressiveImageLoaderModule,
            providers: [
                { provide: IMAGE_LOADER_CONFIG_TOKEN, useValue: config },
                {
                    provide: ConfigurationService,
                    useFactory: ConfigurationServiceFactory,
                    deps: [IMAGE_LOADER_CONFIG_TOKEN]
                }
            ]
        };
    };
    NgxProgressiveImageLoaderModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [
                        ProgressiveImageLoaderComponent,
                        ProgressiveImageDirective,
                        ImagePlaceholderComponent
                    ],
                    exports: [ProgressiveImageLoaderComponent, ProgressiveImageDirective, ImagePlaceholderComponent]
                },] }
    ];
    return NgxProgressiveImageLoaderModule;
}());
export { NgxProgressiveImageLoaderModule };
