// ngx-progressive-image-loader/esm5/lib/image-placeholder/image-placeholder.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { ProgressiveImageLoaderComponent } from '../progressive-image-loader/progressive-image-loader.component';
var ImagePlaceholderComponent = /** @class */ (function () {
    function ImagePlaceholderComponent(sanitizer, _ProgressiveImageLoader) {
        this.sanitizer = sanitizer;
        this._ProgressiveImageLoader = _ProgressiveImageLoader;
        this.class = 'ngx-image-placeholder';
    }
    Object.defineProperty(ImagePlaceholderComponent.prototype, "placeHolder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustStyle("padding-bottom: " + 100 / (this.imageRatio || this._ProgressiveImageLoader.imageRatio) + "%;");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImagePlaceholderComponent.prototype, "imageFilter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustStyle("" + this._ProgressiveImageLoader.filter);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImagePlaceholderComponent.prototype, "safeLoadingImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustUrl(this.placeholderImageSrc || this._ProgressiveImageLoader.placeholderImageSrc);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ImagePlaceholderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    ImagePlaceholderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-image-placeholder',
                    template: "\n    <img class=\"placeholder-loading-image\" [src]=\"safeLoadingImage\" [style.filter]=\"imageFilter\" />\n    <ng-content></ng-content>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    ImagePlaceholderComponent.ctorParameters = function () { return [
        { type: DomSanitizer },
        { type: ProgressiveImageLoaderComponent }
    ]; };
    ImagePlaceholderComponent.propDecorators = {
        class: [{ type: HostBinding, args: ['class',] }],
        placeHolder: [{ type: HostBinding, args: ['style',] }],
        imageRatio: [{ type: Input }],
        placeholderImageSrc: [{ type: Input }]
    };
    return ImagePlaceholderComponent;
}());
export { ImagePlaceholderComponent };
if (false) {
    /** @type {?} */
    ImagePlaceholderComponent.prototype.class;
    /** @type {?} */
    ImagePlaceholderComponent.prototype.imageRatio;
    /** @type {?} */
    ImagePlaceholderComponent.prototype.placeholderImageSrc;
    /**
     * @type {?}
     * @private
     */
    ImagePlaceholderComponent.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    ImagePlaceholderComponent.prototype._ProgressiveImageLoader;
}
