// ngx-progressive-image-loader/fesm5/ngx-progressive-image-loader.js
import { isPlatformBrowser } from '@angular/common';
import { WINDOW } from 'ngx-window-token';
import { __assign } from 'tslib';
import { DomSanitizer } from '@angular/platform-browser';
import { InjectionToken, Inject, Injectable, ChangeDetectionStrategy, Component, Input, Optional, PLATFORM_ID, Renderer2, HostBinding, Directive, ElementRef, EventEmitter, Output, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var IMAGE_LOADER_CONFIG_TOKEN = new InjectionToken('Image loader Config');
/** @type {?} */
var DEFAULT_IMAGE_LOADER_OPTIONS = (/** @type {?} */ ({
    // root?: Element | null;
    rootMargin: '10px',
    threshold: [0.1, 0.5, 1],
    imageRatio: 16 / 9,
    placeholderImageSrc: '',
    isAggressiveLoading: true,
    concurrentLoading: 4
}));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ConfigurationService = /** @class */ (function () {
    function ConfigurationService(options) {
        this.setConfig(options);
    }
    /**
     * @param {?} config
     * @return {?}
     */
    ConfigurationService.prototype.setConfig = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        if (config) {
            this.config = __assign({}, DEFAULT_IMAGE_LOADER_OPTIONS, config);
        }
        else {
            this.config = DEFAULT_IMAGE_LOADER_OPTIONS;
        }
    };
    ConfigurationService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ConfigurationService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [IMAGE_LOADER_CONFIG_TOKEN,] }] }
    ]; };
    return ConfigurationService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} window
 * @return {?}
 */
function isSupportIntersectionObserver(window) {
    return (window &&
        'IntersectionObserver' in window &&
        'IntersectionObserverEntry' in window &&
        'intersectionRatio' in window.IntersectionObserverEntry.prototype);
}
/**
 * @param {?} element
 * @return {?}
 */
function isImagePicture(element) {
    return element instanceof HTMLImageElement || element instanceof HTMLPictureElement;
}
/**
 * @param {?} renderer
 * @param {?} attribute
 * @param {?} element
 * @return {?}
 */
function setAttribute(renderer, attribute, element) {
    renderer.setAttribute(element, attribute, element.dataset[attribute]);
    // maybe doesn't matter
    // renderer.removeAttribute(element, 'data-' + attribute);
}
/**
 * @param {?} element
 * @return {?}
 */
function isPictureElement(element) {
    return element.nodeName === 'PICTURE';
}
/**
 * @param {?} renderer
 * @param {?} image
 * @return {?}
 */
function loadImage(renderer, image) {
    if (isPictureElement(image.parentElement)) {
        /** @type {?} */
        var sourceElms = image.parentElement.children;
        for (var index = 0; index < sourceElms.length; index++) {
            /** @type {?} */
            var element = sourceElms[index];
            if (element instanceof HTMLSourceElement) {
                setAttribute(renderer, 'srcset', element);
            }
            else if (element instanceof HTMLImageElement) {
                setAttribute(renderer, 'src', element);
            }
        }
    }
    else {
        if (image.dataset.src) {
            setAttribute(renderer, 'src', image);
        }
        if (image.dataset.srcset) {
            setAttribute(renderer, 'srcset', image);
        }
    }
}
/**
 * @param {?} window
 * @return {?}
 */
function isSpider(window) {
    return ((window && !('onscroll' in window)) ||
        /(gle|ing|ro)bot|crawl|spider/i.test(window.navigator.userAgent));
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ProgressiveImageLoaderComponent = /** @class */ (function () {
    function ProgressiveImageLoaderComponent(_Renderer, _ConfigurationService, platformId, window) {
        this._Renderer = _Renderer;
        this._ConfigurationService = _ConfigurationService;
        this.platformId = platformId;
        this.window = window;
        // to store observed images
        this.targetMap = new Map();
        // to maintain the sequence of observed images
        this.targetQueue = (/** @type {?} */ ([]));
        // counter of current loading images
        this.loading = 0;
    }
    Object.defineProperty(ProgressiveImageLoaderComponent.prototype, "isObservable", {
        get: /**
         * @return {?}
         */
        function () {
            return !!this.intersectionObserver;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.isAggressiveLoading === undefined) {
            this.isAggressiveLoading = this._ConfigurationService.config.isAggressiveLoading;
        }
        if (this.concurrentLoading === undefined) {
            this.concurrentLoading = this._ConfigurationService.config.concurrentLoading;
        }
        if (this.window &&
            isSupportIntersectionObserver(this.window) &&
            !isSpider(this.window) &&
            isPlatformBrowser(this.platformId)) {
            if (!this.imageRatio) {
                this.imageRatio = this._ConfigurationService.config.imageRatio;
            }
            if (!this.filter) {
                this.filter = this._ConfigurationService.config.filter;
            }
            if (!this.placeholderImageSrc) {
                this.placeholderImageSrc = this._ConfigurationService.config.placeholderImageSrc;
            }
            this.intersectionObserver = new IntersectionObserver(this.onIntersectionChanged.bind(this), this._ConfigurationService.config);
        }
    };
    /**
     * @param {?} target
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.observe = /**
     * @param {?} target
     * @return {?}
     */
    function (target) {
        // so intersection observer can always detect it correctly, otherwise image elements with 0 in height sometime don't load correctly
        target.style.minHeight = '1rem';
        this.intersectionObserver.observe(target);
        this.targetMap.set(target.dataset.src, target);
        this.targetQueue.push(target.dataset.src);
    };
    /**
     * @param {?} target
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.unobserve = /**
     * @param {?} target
     * @return {?}
     */
    function (target) {
        target.style.minHeight = 'initial';
        this.targetMap.delete(target.dataset.src);
        this.intersectionObserver.unobserve(target);
    };
    // called after an image loaded
    // called after an image loaded
    /**
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.imageLoaded = 
    // called after an image loaded
    /**
     * @return {?}
     */
    function () {
        this.loading--;
        while (this.isAggressiveLoading &&
            this.targetQueue &&
            this.targetQueue.length &&
            this.loading <= this.concurrentLoading) {
            /** @type {?} */
            var next = this.targetQueue.shift();
            this.targetMap.has(next) && this.loadImage(this.targetMap.get(next));
        }
    };
    /**
     * @param {?} entries
     * @param {?} observer
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.onIntersectionChanged = /**
     * @param {?} entries
     * @param {?} observer
     * @return {?}
     */
    function (entries, observer) {
        var _this = this;
        entries.forEach((/**
         * @param {?} entry
         * @return {?}
         */
        function (entry) {
            if (entry.isIntersecting) {
                _this.loadImage((/** @type {?} */ (entry.target)));
            }
        }));
    };
    // start loading an image
    // start loading an image
    /**
     * @param {?} image
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.loadImage = 
    // start loading an image
    /**
     * @param {?} image
     * @return {?}
     */
    function (image) {
        // Stop observing the current target
        this.unobserve(image);
        this.loading++;
        loadImage(this._Renderer, image);
    };
    /**
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.targetQueue = [];
        this.targetMap = new Map();
        this.isObservable && this.intersectionObserver.disconnect();
    };
    /**
     * @return {?}
     */
    ProgressiveImageLoaderComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.isObservable && this.intersectionObserver.disconnect();
        this.intersectionObserver = this.targetQueue = this.targetMap = undefined;
    };
    ProgressiveImageLoaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-progressive-image-loader',
                    exportAs: 'ngxProgressiveImageLoader',
                    template: "\n    <ng-content></ng-content>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    ProgressiveImageLoaderComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ConfigurationService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [WINDOW,] }] }
    ]; };
    ProgressiveImageLoaderComponent.propDecorators = {
        imageRatio: [{ type: Input }],
        concurrentLoading: [{ type: Input }],
        isAggressiveLoading: [{ type: Input }],
        filter: [{ type: Input }],
        placeholderImageSrc: [{ type: Input }]
    };
    return ProgressiveImageLoaderComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ImagePlaceholderComponent = /** @class */ (function () {
    function ImagePlaceholderComponent(sanitizer, _ProgressiveImageLoader) {
        this.sanitizer = sanitizer;
        this._ProgressiveImageLoader = _ProgressiveImageLoader;
        this.class = 'ngx-image-placeholder';
    }
    Object.defineProperty(ImagePlaceholderComponent.prototype, "placeHolder", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustStyle("padding-bottom: " + 100 / (this.imageRatio || this._ProgressiveImageLoader.imageRatio) + "%;");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImagePlaceholderComponent.prototype, "imageFilter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustStyle("" + this._ProgressiveImageLoader.filter);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImagePlaceholderComponent.prototype, "safeLoadingImage", {
        get: /**
         * @return {?}
         */
        function () {
            return this.sanitizer.bypassSecurityTrustUrl(this.placeholderImageSrc || this._ProgressiveImageLoader.placeholderImageSrc);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ImagePlaceholderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    ImagePlaceholderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-image-placeholder',
                    template: "\n    <img class=\"placeholder-loading-image\" [src]=\"safeLoadingImage\" [style.filter]=\"imageFilter\" />\n    <ng-content></ng-content>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    ImagePlaceholderComponent.ctorParameters = function () { return [
        { type: DomSanitizer },
        { type: ProgressiveImageLoaderComponent }
    ]; };
    ImagePlaceholderComponent.propDecorators = {
        class: [{ type: HostBinding, args: ['class',] }],
        placeHolder: [{ type: HostBinding, args: ['style',] }],
        imageRatio: [{ type: Input }],
        placeholderImageSrc: [{ type: Input }]
    };
    return ImagePlaceholderComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ProgressiveImageDirective = /** @class */ (function () {
    function ProgressiveImageDirective(_ElementRef, _Renderer, _ConfigurationService, _ImagePlaceholder, _ProgressiveImageLoader) {
        this._ElementRef = _ElementRef;
        this._Renderer = _Renderer;
        this._ConfigurationService = _ConfigurationService;
        this._ImagePlaceholder = _ImagePlaceholder;
        this._ProgressiveImageLoader = _ProgressiveImageLoader;
        this.noPlaceholder = false;
        this.onImageLoaded = new EventEmitter();
        this.isObserved = false;
    }
    Object.defineProperty(ProgressiveImageDirective.prototype, "imageRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this._imageRatio || this._ProgressiveImageLoader.imageRatio;
        },
        // to create a placeholder before finish loading the real image to avoid reflow
        set: 
        // to create a placeholder before finish loading the real image to avoid reflow
        /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._imageRatio = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProgressiveImageDirective.prototype, "placeholderImageSrc", {
        get: /**
         * @return {?}
         */
        function () {
            return this._placeholderImageSrc || this._ProgressiveImageLoader.placeholderImageSrc;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._placeholderImageSrc = value;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ProgressiveImageDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.imageElement = this._ElementRef.nativeElement;
        this.setDataSrc('data-src', this.src);
        this.setDataSrc('data-srcset', this.srcset);
        if (this._ProgressiveImageLoader.isObservable) {
            // only image element need to be observe and have onload event
            if (this.imageElement instanceof HTMLImageElement) {
                this.isObserved = true;
                this._ProgressiveImageLoader.observe(this.imageElement);
                this.imageElement.onload = (/**
                 * @return {?}
                 */
                function () {
                    _this.onImageLoaded.emit(_this.imageElement);
                    _this.imageElement.classList.add('loaded');
                    _this._ProgressiveImageLoader.imageLoaded();
                });
                if (!this._ImagePlaceholder && !this.noPlaceholder) {
                    this.setPlaceholder();
                }
            }
        }
        else {
            // show image directly
            loadImage(this._Renderer, this.imageElement);
            this.imageElement.classList.add('loaded');
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ProgressiveImageDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        changes.src && !changes.src.isFirstChange() && this.setDataSrc('data-src', this.src);
        changes.srcset &&
            !changes.srcset.isFirstChange() &&
            this.setDataSrc('data-srcset', this.srcset);
        if (this.isObserved &&
            ((changes.src && !changes.src.isFirstChange()) ||
                (changes.srcset && !changes.srcset.isFirstChange()))) {
            this._ProgressiveImageLoader.unobserve(this.imageElement);
            this._ProgressiveImageLoader.observe(this.imageElement);
        }
    };
    /**
     * @param {?} attr
     * @param {?} value
     * @return {?}
     */
    ProgressiveImageDirective.prototype.setDataSrc = /**
     * @param {?} attr
     * @param {?} value
     * @return {?}
     */
    function (attr, value) {
        value && this._Renderer.setAttribute(this.imageElement, attr, value);
    };
    /**
     * @return {?}
     */
    ProgressiveImageDirective.prototype.setPlaceholder = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var parentElement = this.imageElement.parentElement;
        /** @type {?} */
        var placeholder = this.createPlaceholder(this.createPlaceholderImage());
        if (isPictureElement(parentElement)) {
            /** @type {?} */
            var pictureParent = parentElement.parentElement;
            this.insertPlaceholder(pictureParent, parentElement, placeholder);
        }
        else {
            this.insertPlaceholder(parentElement, this.imageElement, placeholder);
        }
    };
    /**
     * @private
     * @param {?} parentElement
     * @param {?} imagePicture
     * @param {?} placeholder
     * @return {?}
     */
    ProgressiveImageDirective.prototype.insertPlaceholder = /**
     * @private
     * @param {?} parentElement
     * @param {?} imagePicture
     * @param {?} placeholder
     * @return {?}
     */
    function (parentElement, imagePicture, placeholder) {
        parentElement.insertBefore(placeholder, imagePicture);
        placeholder.style.paddingBottom = 100 / this.imageRatio + "%";
        placeholder.appendChild(imagePicture);
    };
    /**
     * @private
     * @param {?} placeholderImage
     * @return {?}
     */
    ProgressiveImageDirective.prototype.createPlaceholder = /**
     * @private
     * @param {?} placeholderImage
     * @return {?}
     */
    function (placeholderImage) {
        /** @type {?} */
        var placeholder = document.createElement('div');
        placeholder.classList.add('ngx-image-placeholder');
        placeholder.appendChild(placeholderImage);
        return placeholder;
    };
    /**
     * @return {?}
     */
    ProgressiveImageDirective.prototype.createPlaceholderImage = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var img = new Image();
        img.classList.add('placeholder-loading-image');
        img.style.filter = this._ProgressiveImageLoader.filter;
        img.src = this.placeholderImageSrc;
        return img;
    };
    ProgressiveImageDirective.decorators = [
        { type: Directive, args: [{
                    // make sure the element is an image element
                    selector: 'img[ngxProgressiveImage], source[ngxProgressiveImage]'
                },] }
    ];
    /** @nocollapse */
    ProgressiveImageDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ConfigurationService },
        { type: ImagePlaceholderComponent, decorators: [{ type: Optional }, { type: Inject, args: [ImagePlaceholderComponent,] }] },
        { type: ProgressiveImageLoaderComponent, decorators: [{ type: Inject, args: [ProgressiveImageLoaderComponent,] }] }
    ]; };
    ProgressiveImageDirective.propDecorators = {
        imageRatio: [{ type: Input }],
        placeholderImageSrc: [{ type: Input }],
        src: [{ type: Input }],
        srcset: [{ type: Input }],
        noPlaceholder: [{ type: Input }],
        onImageLoaded: [{ type: Output }]
    };
    return ProgressiveImageDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} options
 * @return {?}
 */
function ConfigurationServiceFactory(options) {
    return new ConfigurationService(options);
}
var NgxProgressiveImageLoaderModule = /** @class */ (function () {
    function NgxProgressiveImageLoaderModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    NgxProgressiveImageLoaderModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: NgxProgressiveImageLoaderModule,
            providers: [
                { provide: IMAGE_LOADER_CONFIG_TOKEN, useValue: config },
                {
                    provide: ConfigurationService,
                    useFactory: ConfigurationServiceFactory,
                    deps: [IMAGE_LOADER_CONFIG_TOKEN]
                }
            ]
        };
    };
    NgxProgressiveImageLoaderModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [
                        ProgressiveImageLoaderComponent,
                        ProgressiveImageDirective,
                        ImagePlaceholderComponent
                    ],
                    exports: [ProgressiveImageLoaderComponent, ProgressiveImageDirective, ImagePlaceholderComponent]
                },] }
    ];
    return NgxProgressiveImageLoaderModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { ProgressiveImageLoaderComponent, ProgressiveImageDirective, ImagePlaceholderComponent, ConfigurationServiceFactory, NgxProgressiveImageLoaderModule, IMAGE_LOADER_CONFIG_TOKEN, DEFAULT_IMAGE_LOADER_OPTIONS, isSupportIntersectionObserver, isImagePicture, setAttribute, isPictureElement, loadImage, isSpider, ConfigurationService as ɵa };
